File: Solution/Python/763. Partition Labels.py
class Solution(object):
    def partitionLabels(self, S):
        """
        :type S: str
        :rtype: List[int]
        """
        rangeDict = {}
        for i, c in enumerate(S):
            if c not in rangeDict:
                rangeDict[c] = [i, i]
            else:
                rangeDict[c][1] = i

        rangeList = sorted(rangeDict.values(), key=lambda x: (x[0]))
        ans = []
        cmin = cmax = 0
        for start, end in rangeList:
            if start > cmax:
                ans.append(cmax - cmin + 1)
                cmin, cmax = start, end
            else:
                cmax = max(cmax, end)
        ans.append(cmax - cmin + 1)
        return ans


s = Solution()
res = s.partitionLabels("ababcbacadefegdehijhklij")
print(res)

File: Solution/Javascript/763. Partition Labels.js
/**
 * @param {string} S
 * @return {number[]}
 */
var partitionLabels = function (S) {
    let posDict = {};
    for (let i = 0; i < S.length; i++) {
        let char = S[i];
        if (posDict[char]) {
            posDict[char].end = i;
        } else {
            posDict[char] = { start: i, end: i };
        }
    }
    let res = [];
    let min = max = 0;
    for (let i in posDict) {
        let start = posDict[i].start;
        let end = posDict[i].end;
        if (start > max) {
            res.push(max - min + 1);
            [min, max] = [start, end];
        } else {
            max = Math.max(max, end);
        }
    }
    res.push(max - min + 1);
    return res;
};


let S = "ababcbacadefegdehijhklij";
let res = partitionLabels(S);
console.log(res);